File: subsets2.cpp
class Solution {
    //tc = O(2N*N)- 2n for recursion n for the subset that well be making sc-O(2n) for subsets with average length of o(k) plus O(N) stack space for recusrion
    void solve(vector<vector<int>>& sol, vector<int>& temp, vector<int>& nums, int begin){
        sol.push_back(temp);
        for(int i=begin;i<nums.size();i++){
            if(i>begin && nums[i-1]==nums[i]) continue;
            temp.push_back(nums[i]);
            solve(sol, temp, nums, i+1);
            temp.pop_back();    
        }
    }
    
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> sol;
        sort(nums.begin(),nums.end());
        vector<int> temp;
        solve(sol, temp, nums, 0);
        return sol;
    }
};

File: subsets2.java
class Solution {
    public void solve(int idx, List<List<Integer>> ans, int[] nums, List<Integer> temp){
        ans.add(new ArrayList<>(temp));
        for(int i=idx;i<nums.length; i++){
            if(i!=idx && nums[i]==nums[i-1]){
                continue;
            }else{
                temp.add(nums[i]);
                solve(i+1,ans, nums, temp);
                temp.remove(temp.size()-1);
            }
        }
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        solve(0, ans, nums, new ArrayList<>());
        return ans;
    }
}